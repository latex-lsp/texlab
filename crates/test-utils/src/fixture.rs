use std::path::PathBuf;

use base_db::{util::LineCol, Owner, Workspace};
use rowan::{TextRange, TextSize};
use url::Url;

#[derive(Debug)]
pub struct Fixture {
    pub workspace: Workspace,
    pub documents: Vec<DocumentSpec>,
}

impl Fixture {
    pub fn parse(input: &str) -> Fixture {
        let mut documents = Vec::new();

        let mut start = 0;
        for end in input
            .match_indices("%!")
            .skip(1)
            .map(|(i, _)| i)
            .chain(std::iter::once(input.len()))
        {
            documents.push(DocumentSpec::parse(&input[start..end]));
            start = end;
        }

        let mut workspace = Workspace::default();
        for document in &documents {
            let path = PathBuf::from(document.uri.path());
            let language = distro::Language::from_path(&path).unwrap_or(distro::Language::Tex);

            workspace.open(
                document.uri.clone(),
                document.text.clone(),
                language,
                Owner::Client,
                LineCol { line: 0, col: 0 },
            );
        }

        Self {
            workspace,
            documents,
        }
    }
}

#[derive(Debug)]
pub struct DocumentSpec {
    pub uri: Url,
    pub text: String,
    pub cursor: Option<TextSize>,
    pub ranges: Vec<TextRange>,
}

impl DocumentSpec {
    pub fn parse(input: &str) -> Self {
        let (uri_str, input) = input
            .trim()
            .strip_prefix("%! ")
            .map(|input| input.split_once('\n').unwrap_or((input, "")))
            .unwrap();

        let uri = Url::parse(&format!("file:///texlab/{uri_str}")).unwrap();

        let mut ranges = Vec::new();
        let mut cursor = None;

        let mut text = String::new();
        for line in input.lines().map(|line| line.trim_end()) {
            if line.chars().all(|c| matches!(c, ' ' | '^' | '|' | '!')) && !line.is_empty() {
                cursor = cursor.or_else(|| {
                    let offset = line.find('|')?;
                    Some(TextSize::from((text.len() + offset) as u32))
                });

                if let Some(start) = line.find('!') {
                    let position = TextSize::from((text.len() + start) as u32);
                    ranges.push(TextRange::new(position, position));
                }

                if let Some(start) = line.find('^') {
                    let end = line.rfind('^').unwrap() + 1;
                    ranges.push(TextRange::new(
                        TextSize::from((text.len() + start) as u32),
                        TextSize::from((text.len() + end) as u32),
                    ));
                }
            } else {
                text.push_str(line);
                text.push('\n');
            }
        }

        Self {
            uri,
            text,
            cursor,
            ranges,
        }
    }
}
