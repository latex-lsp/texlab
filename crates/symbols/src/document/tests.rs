use base_db::{Config, SymbolConfig};
use insta::assert_debug_snapshot;
use regex::Regex;
use test_utils::fixture::Fixture;

use crate::document_symbols;

#[test]
fn test_enumerate() {
    let fixture = Fixture::parse(
        r#"
%! main.tex
\documentclass{article}

\begin{document}

\begin{enumerate}
    \item\label{it:foo} Foo
    \item\label{it:bar} Bar
    \item[Baz] Baz
    \item[Qux]\label{it:qux} Qux
\end{enumerate}

\end{document}

%! main.aux
\relax
\newlabel{it:foo}{{1}{1}}
\newlabel{it:qux}{{2}{1}}"#,
    );

    let document = fixture.workspace.lookup(&fixture.documents[0].uri).unwrap();
    assert_debug_snapshot!(document_symbols(&fixture.workspace, document));
}

#[test]
fn test_equation() {
    let fixture = Fixture::parse(
        r#"
%! main.tex
\documentclass{article}

\begin{document}

\begin{equation}\label{eq:foo}
    Foo
\end{equation}

\begin{equation}\label{eq:bar}
    Bar
\end{equation}

\begin{equation}
    Baz
\end{equation}

\end{document}

%! main.aux
\relax
\newlabel{eq:foo}{{1}{1}}"#,
    );

    let document = fixture.workspace.lookup(&fixture.documents[0].uri).unwrap();
    assert_debug_snapshot!(document_symbols(&fixture.workspace, document));
}

#[test]
fn test_float() {
    let fixture = Fixture::parse(
        r#"
%! main.tex
\documentclass{article}

\begin{document}

\begin{figure}
    Foo
    \caption{Foo}\label{fig:foo}
\end{figure}

\begin{figure}
    Bar
    \caption{Bar}\label{fig:bar}
\end{figure}

\begin{figure}
    Baz
    \caption{Baz}
\end{figure}

\begin{figure}
    Qux
\end{figure}

\end{document}
|

%! main.aux
\relax
\@writefile{lof}{\contentsline {figure}{\numberline {1}{\ignorespaces Foo}}{1}\protected@file@percent }
\newlabel{fig:foo}{{1}{1}}
\@writefile{lof}{\contentsline {figure}{\numberline {2}{\ignorespaces Bar}}{1}\protected@file@percent }
\@writefile{lof}{\contentsline {figure}{\numberline {3}{\ignorespaces Baz}}{1}\protected@file@percent }"#,
    );

    let document = fixture.workspace.lookup(&fixture.documents[0].uri).unwrap();
    assert_debug_snapshot!(document_symbols(&fixture.workspace, document));
}

#[test]
fn test_section() {
    let fixture = Fixture::parse(
        r#"
%! main.tex
\documentclass{article}

\begin{document}

\section{Foo}

\section{Bar}\label{sec:bar}

\subsection{Baz}\label{sec:baz}

\end{document}
|

%! main.aux
\relax
\@writefile{toc}{\contentsline {section}{\numberline {1}Foo}{1}\protected@file@percent }
\@writefile{toc}{\contentsline {section}{\numberline {2}Bar}{1}\protected@file@percent }
\newlabel{sec:bar}{{2}{1}}"#,
    );

    let document = fixture.workspace.lookup(&fixture.documents[0].uri).unwrap();
    assert_debug_snapshot!(document_symbols(&fixture.workspace, document));
}

#[test]
fn test_theorem_amsthm() {
    let fixture = Fixture::parse(
        r#"
%! main.tex
\documentclass{article}
\usepackage{amsthm}
\newtheorem{lemma}{Lemma}

\begin{document}

\begin{lemma}[Foo]\label{thm:foo}
    Foo
\end{lemma}

\begin{lemma}\label{thm:bar}
    Bar
\end{lemma}

\begin{lemma}\label{thm:baz}
    Baz
\end{lemma}

\begin{lemma}[Qux]
    Qux
\end{lemma}

\end{document}
|

%! main.aux
\relax
\newlabel{thm:foo}{{1}{1}}
\newlabel{thm:bar}{{2}{1}}"#,
    );

    let document = fixture.workspace.lookup(&fixture.documents[0].uri).unwrap();
    assert_debug_snapshot!(document_symbols(&fixture.workspace, document));
}

#[test]
fn test_theorem_thmtools() {
    let fixture = Fixture::parse(
        r#"
%! main.tex
\documentclass{article}
\declaretheoremstyle{lemmastyle}
\declaretheorem[style=lemmastyle, name=Lemma]{lemma}

\begin{document}

\begin{lemma}[Foo]\label{thm:foo}
    Foo
\end{lemma}

\begin{lemma}\label{thm:bar}
    Bar
\end{lemma}

\begin{lemma}\label{thm:baz}
    Baz
\end{lemma}

\begin{lemma}[Qux]
    Qux
\end{lemma}

\end{document}
|

%! main.aux
\relax
\newlabel{thm:foo}{{1}{1}}
\newlabel{thm:bar}{{2}{1}}"#,
    );

    let document = fixture.workspace.lookup(&fixture.documents[0].uri).unwrap();
    assert_debug_snapshot!(document_symbols(&fixture.workspace, document));
}

#[test]
fn test_allowed_patterns() {
    let mut fixture = Fixture::parse(
        r#"
%! main.tex
\documentclass{article}

\begin{document}

\begin{equation}\label{eq:foo}
    Foo
\end{equation}

\begin{enumerate}
    \item Foo
    \item Bar
\end{enumerate}

\end{document}"#,
    );

    fixture.workspace.set_config(Config {
        symbols: SymbolConfig {
            allowed_patterns: vec![
                Regex::new("Item").unwrap(),
                Regex::new("Enumerate").unwrap(),
            ],
            ignored_patterns: Vec::new(),
        },
        ..Config::default()
    });

    let document = fixture.workspace.lookup(&fixture.documents[0].uri).unwrap();
    assert_debug_snapshot!(document_symbols(&fixture.workspace, document));
}

#[test]
fn test_ignored_patterns() {
    let mut fixture = Fixture::parse(
        r#"
%! main.tex
\documentclass{article}

\begin{document}

\begin{equation}\label{eq:foo}
    Foo
\end{equation}

\begin{enumerate}
    \item Foo
    \item Bar
\end{enumerate}

\end{document}"#,
    );

    fixture.workspace.set_config(Config {
        symbols: SymbolConfig {
            ignored_patterns: vec![
                Regex::new("Item").unwrap(),
                Regex::new("Enumerate").unwrap(),
            ],
            allowed_patterns: Vec::new(),
        },
        ..Config::default()
    });

    let document = fixture.workspace.lookup(&fixture.documents[0].uri).unwrap();
    assert_debug_snapshot!(document_symbols(&fixture.workspace, document));
}
