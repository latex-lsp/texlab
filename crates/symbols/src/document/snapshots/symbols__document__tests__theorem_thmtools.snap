---
source: crates/symbols/src/document/tests.rs
expression: "document_symbols(&fixture.workspace, document)"
---
[
    Symbol {
        name: "Lemma 1 (Foo)",
        kind: Theorem,
        label: Some(
            Span(
                "thm:foo",
                147..162,
            ),
        ),
        full_range: 129..182,
        selection_range: 147..162,
        children: [],
    },
    Symbol {
        name: "Lemma 2",
        kind: Theorem,
        label: Some(
            Span(
                "thm:bar",
                197..212,
            ),
        ),
        full_range: 184..232,
        selection_range: 197..212,
        children: [],
    },
    Symbol {
        name: "Lemma",
        kind: Theorem,
        label: Some(
            Span(
                "thm:baz",
                247..262,
            ),
        ),
        full_range: 234..282,
        selection_range: 247..262,
        children: [],
    },
    Symbol {
        name: "Lemma (Qux)",
        kind: Theorem,
        label: None,
        full_range: 284..322,
        selection_range: 284..322,
        children: [],
    },
]
